Commonize `:compose:material3:adaptive` modules

terrakok opened the pull request:
Commonize and configure publishing for 
 - `:compose:material3:adaptive:adaptive`
 - `:compose:material3:adaptive:adaptive-layout`
 - `:compose:material3:adaptive:adaptive-navigation`
 - `:window:window-core`
modules.

Added new method to non-android source set to get `WindowAdaptiveInfo`. 

Fixes https://youtrack.jetbrains.com/issue/CMP-5817

## Testing
Checked it in the MPP demo app:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/e2dbe01f-1481-4e4f-ab9d-e2641bc441d0">



## Release Notes
### Features - Multiple Platforms
- Commonized `:compose:material3:adaptive:adaptive`, `:compose:material3:adaptive:adaptive-layout`, `:compose:material3:adaptive:adaptive-navigation` and `:window:window-core` modules

igordmn commented:
> Just copy-pasted material3/adaptive directory from the androidx to JetBrains fork

Please specify the source commit. is it `androidx/compose-material3/1.3.0-beta03`?

igordmn commented on compose/material3/adaptive/adaptive-layout/src/jvmStubsMain/kotlin/androidx/compose/material3/adaptive/layout/NotImplemented.jvmStubs.kt:
```diff
@@ -0,0 +1,27 @@
+package androidx.compose.material3.adaptive.layout
+
+@Suppress("NOTHING_TO_INLINE")
+internal inline fun implementedInJetBrainsFork(): Nothing =
```
This shouldn't be in `androidx/compose-material3/1.3.0-beta03`, the stubs still weren't merged to `integration` or `jb-main`

terrakok committed:
Commonize `:window:window-core` module and configure publication.

terrakok committed:
Commonize `:compose:material3:adaptive:adaptive` module and configure publication.

terrakok committed:
Add an usage of `material3:adaptive:adaptive` to the MPP demo

terrakok committed:
Commonize `:compose:material3:adaptive:adaptive-layout` module and configure publication.

terrakok committed:
Commonize `:compose:material3:adaptive:adaptive-navigation` module and configure publication.

MatkovIvan commented on compose/material3/adaptive/adaptive-layout/src/commonTest/kotlin/androidx/compose/material3/adaptive/layout/PaneScaffoldDirectiveTest.kt:
```diff
@@ -38,11 +36,11 @@ class PaneScaffoldDirectiveTest {
                 WindowAdaptiveInfo(WindowSizeClass.compute(400f, 800f), Posture())
             )
 
-        assertThat(scaffoldDirective.maxHorizontalPartitions).isEqualTo(1)
```
It's ported in `:kruth` module, no need to rewrite every check

terrakok replied:
updated

terrakok committed:
Restore original `:compose:material3:adaptive:adaptive-layout` test checks.

MatkovIvan commented on MULTIPLATFORM.md:
```diff
@@ -71,7 +71,7 @@ Compose Multiplatform core libraries can be published to local Maven with the fo
 
 And library groups:
-`-Pjetbrains.publication.libraries=CORE_BUNDLE,COMPOSE,LIFECYCLE,NAVIGATION,SAVEDSTATE`
+`-Pjetbrains.publication.libraries=CORE_BUNDLE,COMPOSE,LIFECYCLE,NAVIGATION,SAVEDSTATE,WINDOW`
```
Do we need to adopt our CI for it?

terrakok replied:
No. By default we publish all libraries

MatkovIvan replied:
As far as I remember, we still need to specify a default version there. cc @igordmn

igordmn replied:
Yes, we need, but I'll done it myself before the release, see my other comment, we can't enable publication just now

MatkovIvan commented on compose/material3/adaptive/adaptive-navigation/build.gradle:
I'm surprised to see no dependencies to a navigation library here.

terrakok replied:
It is quite small module with independent logic how to behave when a screen has different layouts

MatkovIvan commented on libraryversions.toml:
```diff
@@ -292,6 +292,7 @@ WEAR_TILES = { group = "androidx.wear.tiles", atomicGroupVersion = "versions.WEA
 WEAR_WATCHFACE = { group = "androidx.wear.watchface", atomicGroupVersion = "versions.WEAR_WATCHFACE" }
 WEBKIT = { group = "androidx.webkit", atomicGroupVersion = "versions.WEBKIT" }
 WINDOW = { group = "androidx.window", atomicGroupVersion = "versions.WINDOW" }
+WINDOW_CORE = { group = "org.jetbrains.window.core", atomicGroupVersion = "versions.WINDOW", overrideInclude = [ ":window:window-core" ]  }
```
"versions.WINDOW" is already `1.6.0-alpha01` in AOSP. Shouldn't we update it here?

terrakok replied:
I didn't touch the version. As I understand the version was set in the commit we based our fork on

igordmn replied:
This is because we don't merge `window` and `libversions` to jb-main, they can be old. The actual state is in `integration`, but as with any other lib we merge only released commits to `jb-main`

MatkovIvan commented on gradle.properties:
```diff
@@ -108,12 +108,16 @@ artifactRedirecting.androidx.compose.foundation.version=1.7.0-beta05
 artifactRedirecting.androidx.navigation.version=2.8.0-beta05
 artifactRedirecting.androidx.savedstate.version=1.2.1
+# Look for `WINDOW` in libraryversions.toml
+artifactRedirecting.androidx.window.core.version=1.3.0-alpha01
```
The stable 1.3.0 was release a few months ago. Any reason to use alpha?

terrakok replied:
I didn't touch the version. As I understand the version was set in the commit we based our fork on

MatkovIvan replied:
window subfolder wasn't involved in specific version selection, If it's outdated, please merge the stable state (might be done separately)

MatkovIvan replied:
Actually, I'm almost sure that this version doesn't match to the commit/state of the code. Blocking the merge before explicit check

MatkovIvan reviewed (approved):
LGTM overall, no blockers. Please recheck the versions on our CI and redirecting artifacts before merge

terrakok committed:
Change `androidx.window.core` redirect version to `1.3.0-beta02`.

It was pinned in the original `androidx.compose.material3.adaptive` module dependencies

igordmn commented on compose/material3/adaptive/adaptive-layout/src/commonTest/kotlin/androidx/compose/material3/adaptive/layout/ThreePaneMotionTest.kt:
```diff
@@ -320,11 +318,11 @@ class ThreePaneMotionTest {
-                    visibilityThreshold = 0.1f,
-                    delayedRatio = delayedRatio,
-                )
+                dampingRatio = 0.7f,
```
Let's reset formatting to avoid merge conflicts (in the other files too)

igordmn commented on gradle.properties:
```diff
@@ -108,12 +108,15 @@ artifactRedirecting.androidx.compose.foundation.version=1.7.0-beta05
 # Look for `COMPOSE_MATERIAL3_COMMON` in libraryversions.toml
 artifactRedirecting.androidx.compose.material3.common.version=1.0.0-alpha01
+# Look for `COMPOSE_MATERIAL3_ADAPTIVE` in libraryversions.toml
+artifactRedirecting.androidx.compose.material3.adaptive.version=1.0.0-alpha01
```
We didn't take into account that material3 adaptive has own release cycle.

Window-core has a separate release cycle as well.

Because of that we need:
- find a commit for the latest version
- partially merge subfolders of this commit to jb-main.

Please, create tasks on me:
- merge the latest material3.adaptive
- merge the latest androidx.window.core
- enable publication of material3.adaptive/androidx.window.core

Until we merge them, these specified versions can be unaligned with the actual state of the source, and we can't publish.

Let's remove also changes in `mpp/build.gradle.kts`, then we can merge the PR in the current state.

igordmn replied:
>  then we can merge the PR in the current state.

Actually it might be better to merge it after, but I am not sure when I able to merge these folders.

terrakok committed:
Restore files formatting.

terrakok committed:
Fix PaneMotionTest imports

terrakok committed:
Dump `:compose:material3:adaptive:adaptive-layout` and `:compose:material3:adaptive:adaptive-navigation` API.

igordmn commented on gradle.properties:
```diff
@@ -108,12 +108,15 @@ artifactRedirecting.androidx.compose.foundation.version=1.7.0-beta06
 # Look for `COMPOSE_MATERIAL3_COMMON` in libraryversions.toml
 artifactRedirecting.androidx.compose.material3.common.version=1.0.0-alpha01
+# Look for `COMPOSE_MATERIAL3_ADAPTIVE` in libraryversions.toml
+artifactRedirecting.androidx.compose.material3.adaptive.version=1.0.0-alpha01
```
```suggestion
artifactRedirecting.androidx.compose.material3.adaptive.version=1.0.0-beta04
```

igordmn commented on gradle.properties:
```diff
@@ -108,12 +108,15 @@ artifactRedirecting.androidx.compose.foundation.version=1.7.0-beta06
 artifactRedirecting.androidx.lifecycle.version=2.8.0
 artifactRedirecting.androidx.navigation.version=2.8.0-beta05
 artifactRedirecting.androidx.savedstate.version=1.2.1
+artifactRedirecting.androidx.window.core.version=1.3.0-beta02
```
```suggestion
artifactRedirecting.androidx.window.core.version=1.3.0
```

terrakok committed:
Update redirecting versions.

terrakok committed:
Fix import.
